sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/BulkExecutorUtil.java

```diff
@@ -0,0 +1,149 @@
+
+        if (result.getRetryAfterDuration() != null) {
+            headers.put(HttpConstants.HttpHeaders.RETRY_AFTER_IN_MILLISECONDS, String.valueOf(result.getRetryAfterDuration().toMillis()));
+        }
+
+        return headers;
+    }
+
+    // TODO: metaDataDiagnosticContext is passed as null. Better way is to add this context to the cosmos diagnostic.
+    static Mono<String> resolvePartitionKeyRangeId(
+        AsyncDocumentClient docClientWrapper,
+        CosmosAsyncContainer container,
+        CosmosItemOperation operation) {
+
+        checkNotNull(operation, "expected non-null operation");
+
+        if(operation instanceof ItemBulkOperation<?>) {
+            final ItemBulkOperation<?> itemBulkOperation = (ItemBulkOperation<?>) operation;
+
+            final Mono<String> pkRangeIdMono = BulkExecutorUtil.getCollectionInfoAsync(docClientWrapper, container)
+                .flatMap(collection -> {
+                    final PartitionKeyDefinition definition = collection.getPartitionKey();
+                    final PartitionKeyInternal partitionKeyInternal = getPartitionKeyInternal(operation, definition);
+                    itemBulkOperation.setPartitionKeyJson(partitionKeyInternal.toJson());
+
+                    return docClientWrapper.getPartitionKeyRangeCache()
+                        .tryLookupAsync(null, collection.getResourceId(), null, null)
+                        .map((Utils.ValueHolder<CollectionRoutingMap> routingMap) -> {
+
+                            return routingMap.v.getRangeByEffectivePartitionKey(
+                                getEffectivePartitionKeyString(
+                                    partitionKeyInternal,
+                                    definition)).getId();
+                        });
+                });
+
+            return pkRangeIdMono;
+        } else {
+            throw new UnsupportedOperationException("Unknown CosmosItemOperation.");
+        }
+    }
+
+    private static PartitionKeyInternal getPartitionKeyInternal(
+        final CosmosItemOperation operation,
+        final PartitionKeyDefinition partitionKeyDefinition) {
+
+        checkNotNull(operation, "expected non-null operation");
+
+        final PartitionKey partitionKey = operation.getPartitionKeyValue();
+        if (partitionKey == null) {
+            return ModelBridgeInternal.getNonePartitionKey(partitionKeyDefinition);
+        } else {
+            return BridgeInternal.getPartitionKeyInternal(partitionKey);
+        }
+    }
+
+    private static Mono<DocumentCollection> getCollectionInfoAsync(AsyncDocumentClient documentClient,
+                                                                   CosmosAsyncContainer container) {
+        final RxClientCollectionCache clientCollectionCache = documentClient.getCollectionCache();
+        return clientCollectionCache
+            .resolveByNameAsync(
+                null,
```

Comment:
doesn't passing null, result in missing to add this step to request diagnostics?

Reply:
I still see that this hasn't been done. could you please add a TODO in the code with your alias?

Reply:
Apologise Mo. Will add one. This all should be handled when we are adding diagnostic operation wise.